modepy/matrices.py

```diff
@@ -351,11 +355,131 @@ def mass_matrix(
+        proj_functions: Sequence[Callable[[np.ndarray], np.ndarray]],
+        nodes: np.ndarray | None = None
+    ) -> np.ndarray:
+    r"""Using *quadrature*, provide a matrix :math:`A` that satisfies:
+
+    .. math::
+
+        \displaystyle (A \boldsymbol u)_i = \sum_j w_j \phi_i(r_j) u_j,
+
+    where :math:`\phi_i` are the Lagrange basis functions obtained from
+    *test_functions* at *nodes*, :math:`w_j` and :math:`r_j` are the weights
+    and nodes from *quadrature*, and :math:`u_j` are point values of a trial
+    solution at the *quadrature* nodes.
+    """
+    if nodes is None:
+        nodes = quadrature.nodes
+
+    if len(proj_functions) != nodes.shape[1]:
+        raise ValueError("volume_nodes not unisolvent with proj_functions")
+
+    vdm = vandermonde(proj_functions, nodes)
+
+    return la.solve(
+        vdm.T, modal_quad_operator(
+            quadrature,
+            test_functions
+        )
+    )
+
+
+def modal_quad_bilinear_form(
+            quadrature: Quadrature,
+            test_functions: Sequence[Callable[[np.ndarray], np.ndarray]],
+            trial_functions: Sequence[Callable[[np.ndarray], np.ndarray]],
+        ) -> np.ndarray:
+    r"""Using *quadrature*, provide a matrix :math:`A` defined as:
+
+    .. math::
+
+        \displaystyle A_{ij} = \sum_k \psi_i(r_k) \phi_j(r_k) w_k,
+
+    where :math:`\psi_i` are from *test_functions*, :math:`\phi_j` are from
+    *trial_functions*, :math:`r_k` and :math:`w_k` are nodes and weights
+    from *quadrature*. The matrix :math:`A` satisfies
+
+    .. math::
+
+        \displaystyle (u, \psi_i)_A = \sum_{j} A_{ij} u_j,
+
+    where :math:`u_i` are modal coefficients of a trial solution.
+    """
+    return np.einsum(
+        "qi,qj,q->ij",
+        vandermonde(test_functions, quadrature.nodes),
+        vandermonde(trial_functions, quadrature.nodes),
+        quadrature.weights
+    )
+
+
+def nodal_quad_bilinear_form(
+            quadrature: Quadrature,
+            test_functions: Sequence[Callable[[np.ndarray], np.ndarray]],
+            trial_functions: Sequence[Callable[[np.ndarray], np.ndarray]],
+            proj_functions: Sequence[Callable[[np.ndarray], np.ndarray]],
+            nodes: np.ndarray
+        ) -> np.ndarray:
+    r"""Using *quadrature*, provide a matrix :math:`A` defined as:
+
+    .. math::
+
+        \displaystyle A_{ij} = \sum_k \psi_i(r_k) \phi_j(r_k) w_k,
+
+    where :math:`\psi_i` are Lagrange basis functions obtained from
+    *test_functions*, :math:`\phi_j` are Lagrange basis functions obtained from
+    *trial_functions*, :math:`r_k` and :math:`w_k` are nodes and weights from
+    *quadrature*. The matrix :math:`A` satisfies
+
+    .. math::
+
+        \displaystyle (u, \psi_i)_A = \sum_{j} A_{ij} u_j,
+
+    where :math:`u_i` are nodal coefficients of a trial solution.
+    """
+    if len(test_functions) != nodes.shape[1]:
+        raise ValueError("volume_nodes not unisolvent with proj_functions")
```

Comment:
```suggestion
    if len(proj_functions) != nodes.shape[1]:
        raise ValueError("volume_nodes not unisolvent with proj_functions")
```
(but this is probably going away anyway)